quinn-proto/src/connection/mod.rs

```diff
@@ -2376,25 +2529,51 @@ where
 
+                    let mut valid_rem_cid = Ok(());
+                    if frame.retire_prior_to > self.rem_cid_seq {
+                        // If our current CID is earlier than the first unretired one we must not
+                        // have adopted the one we just got, so it must be stored in rem_cids, so
+                        // this unwrap is guaranteed to succeed.
+                        valid_rem_cid = self.update_rem_cid();
+                    }
+
                     // Must be after retire_prior_to processing in case this CID has the highest
                     // permitted sequence number
                     use crate::cid_queue::InsertError;
-                    match self.rem_cids.insert(IssuedCid {
+                    let mut new_rem_cid = IssuedCid {
                         sequence: frame.sequence,
                         id: frame.id,
                         reset_token: frame.reset_token,
-                    }) {
-                        Ok(()) => {}
-                        Err(InsertError::ExceedsLimit) => {
-                            return Err(TransportError::CONNECTION_ID_LIMIT_ERROR(""));
-                        }
-                        Err(InsertError::Retired) => {
-                            trace!("discarding already-retired");
-                            self.spaces[SpaceId::Data]
-                                .pending
-                                .retire_cids
-                                .push(frame.sequence);
-                            continue;
+                    };
+
+                    // Guarantee the cid sequence in active rem cid (self.rem_cid) is always smallest among active ones
+                    if valid_rem_cid.is_ok() && frame.sequence + 1 == self.rem_cids.offset {
+                        let swap_cid = new_rem_cid;
+                        new_rem_cid.sequence = self.rem_cid_seq;
+                        new_rem_cid.id = self.rem_cid;
+                        new_rem_cid.reset_token = self.rem_cid_resettoken.unwrap();
+                        self.apply_new_rem_cid(swap_cid);
+                    }
+
+                    match valid_rem_cid {
+                        Ok(_) => match self.rem_cids.insert(new_rem_cid) {
+                            Ok(()) => {}
+                            Err(InsertError::ExceedsLimit) => {
+                                return Err(TransportError::CONNECTION_ID_LIMIT_ERROR(""));
+                            }
+                            Err(InsertError::Retired) => {
+                                trace!("discarding already-retired");
+                                self.spaces[SpaceId::Data]
+                                    .pending
+                                    .retire_cids
+                                    .push(frame.sequence);
+                                continue;
+                            }
+                        },
+                        Err(_) => {
```

Comment:
If we ignore the `Err` payload, `valid_rem_cid` should just be an `Option`.

Reply:
Will update the corresponding function.